wit-0.3.0-draft/terminal.wit

```diff
@@ -1,25 +1,358 @@
+        /// | `␛[?25l`    | Set the cursor as invisible                       | cursor_invisible | civis | vi |
+        /// | `␛[?1049l`  | Exit full-screen mode and restore the terminal to its prior state | exit_ca_mode | rmcup | te |
+        /// | `␛[?2004l`  | End bracketed paste mode                          | BD        | BD    | \* |
+        /// | `␛[!p`      | Reset the terminal to default settings, without clearing the screen | \* | \* | \* |
+        full-screen-supported: func() -> bool;
+    }
+
+    /// Flags indicating support for different sets of color escape
+    /// sequences, and the user's preference for whether they should
+    /// be used by default.
+    ///
+    /// Other color features, including 88-color and 256-color are not
+    /// included here, as the associated escape sequences are not as
+    /// portable, and they're effectively obviated by truecolor support.
+    flags color-flags {
+        /// Are the classic "4-bit color" escape sequences supported?
+        ///
+        /// This indicates support for 16 colors, on foreground and background,
+        /// using the widely-supported (and ECMA-48) "SGR" color escape
+        /// sequences of the form `␛[…m`. See [here] for more information.
+        ///
+        /// Before using color in your user interface, also consider
+        /// checking `color-desired` to obtain the user's preference for
+        /// enabling color by default.
+        ///
+        /// [here]: https://en.wikipedia.org/wiki/ANSI_escape_code#3-bit_and_4-bit
+        ansi,
+
+        /// Are the 24-bit "true color" escape sequences supported?
+        ///
+        /// This indicates support over 16 million colors, on foreground and
+        /// background, using "true color" escape sequences of the form
+        /// `␛[38;2;«r»;«g»;«b»m` (foreground) and `␛[48;2;«r»;«g»;«b»m`
+        /// (background). The `«r»`, `«g»`, and `«b»` fields are integers
+        /// in the range [0,256) indicating red, green, and blue values
+        /// respectively. See [here] for more information.
+        ///
+        /// Before using color in your user interface, also consider
+        /// checking `color-desired` to obtain the user's preference for
+        /// enabling color by default.
+        ///
+        /// [here]: https://en.wikipedia.org/wiki/ANSI_escape_code#24-bit
+        truecolor,
+
+        /// Does the user want color to be used by default?
+        ///
+        /// Some users have a terminal which supports color, but prefer
+        /// applications not use it by default; this flag indicates
+        /// this preference.
+        ///
+        /// See the [`NO_COLOR` website](http://no-color.org/) for more
+        /// information.
+        color-desired-by-default,
+    }
+
+    /// A pair of rows and columns.
+    record rows-and-columns {
+        rows: u16,
+        columns: u16,
+    }
 }
 
 /// An interface providing an optional `terminal-input` for stdin as a
```

Comment:
should there also be an interface that's equivalent to opening `CON` or `/dev/tty`? this is used by things like `sudo tee` where the inputs/outputs may be redirected but `sudo` still accesses the controlling terminal to write out the password prompt and read in the password.

Reply:
I think it would be straightforward to have an API that returns a `terminal-input`. `terminal-output`, and streams for the controlling terminal (alongside the stdio functions). That's independent of what the `terminal-input` and `terminal-output` interfaces look like though, so that could be proposed separately.

A potential consideration is that we may not always want sandboxed programs to be able to do that, because that would mean they could bypass stdio redirects and potentially grab input from the terminal meant for another application. So we'd want to think through the use cases.